### PR TITLE
Updated GCC/Python Support

### PR DESCRIPTION
**Pull request by Myoldmopar**
On a new machine running Ubuntu 23.10, I encountered several issues.  Part of them due to the update to GCC 13.2, part due to updating to Python 3.11/3.12, and part due to updated Clang Format. I'll provide a walkthrough of the changes, and hopefully they can be massaged into a nice set of changes that work for both our release versions as well as my dev machine.

### CODE REVIEW

File: cmake/CompilerFlags.cmake [hunk 1]
```diff
@@ -86,9 +86,10 @@ elseif(CMAKE_COMPILER_IS_GNUCXX OR "${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang" O
 
   # COMPILER FLAGS
   target_compile_options(project_options INTERFACE -pipe) # Faster compiler processing
-  target_compile_options(project_warnings INTERFACE -Wpedantic
-  )# Turn on warnings about constructs/situations that may be non-portable or outside of the standard
-  target_compile_options(project_warnings INTERFACE -Wall -Wextra) # Turn on warnings
+  target_compile_options(project_warnings INTERFACE -Wpedantic)
+  # Turn on warnings about constructs/situations that may be non-portable or outside of the standard
+  target_compile_options(project_warnings INTERFACE -Wall) # Turn on warnings
+  target_compile_options(project_warnings INTERFACE -Wextra) # Turn on warnings
```
Comment (Myoldmopar) [inline cmake/CompilerFlags.cmake:92]: No functional change, just cleaning up these lines

File: cmake/CompilerFlags.cmake [hunk 2]
```diff
@@ -101,6 +102,11 @@ elseif(CMAKE_COMPILER_IS_GNUCXX OR "${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang" O
     target_compile_options(project_warnings INTERFACE -Wno-unused-but-set-parameter -Wno-unused-but-set-variable)
     target_compile_options(project_warnings INTERFACE -Wno-maybe-uninitialized)
     target_compile_options(project_warnings INTERFACE -Wno-aggressive-loop-optimizations)
+    # Sadly, GCC 13.2 is throwing many false positives on dangling references and compile time array-bounds
+    # https://gcc.gnu.org/git/gitweb.cgi?p=gcc.git;h=6b927b1297e66e26e62e722bf15c921dcbbd25b9
+    target_compile_options(project_warnings INTERFACE -Wno-dangling-reference)
+    target_compile_options(project_warnings INTERFACE -Wno-array-bounds)
+    target_compile_options(project_warnings INTERFACE -Wno-stringop-overflow)
```
Comment (Myoldmopar) [inline cmake/CompilerFlags.cmake:114]: Alright, so with GCC 13.2, I am getting lots of what appear to be false-positives.  I'm not the only one seeing the warnings unexpectedly based on internet searches.  Before this merges, I would like this to be investigated further, and:
 - understand why the errors are being emitted, are they **actually** false positives?
 - be more surgical with the warning mutes, if they will stay at any level

File: src/ConvertInputFormat/main.cpp [hunk 3]
```diff
@@ -88,7 +88,7 @@ static constexpr std::array<std::string_view, static_cast<int>(OutputTypes::Num)
 
 static constexpr auto outputTypeExperimentalStart = OutputTypes::CBOR;
 
-template <typename... Args> void displayMessage(std::string_view str_format, Args &&... args)
+template <typename... Args> void displayMessage(std::string_view str_format, Args &&...args)
```
Comment (Myoldmopar) [inline src/ConvertInputFormat/main.cpp]: Format only

File: src/EnergyPlus/EPVector.hh [hunk 4]
```diff
@@ -263,12 +263,12 @@ private:
     bool m_allocated{false};
 };
 
-template <typename T>[[nodiscard]] bool allocated(EPVector<T> const &v) noexcept
+template <typename T> [[nodiscard]] bool allocated(EPVector<T> const &v) noexcept
```
Comment (Myoldmopar) [inline src/EnergyPlus/EPVector.hh]: Format only in this file.

File: src/EnergyPlus/FileSystem.hh [hunk 5]
```diff
@@ -229,8 +229,8 @@ namespace FileSystem {
         is_any<T, std::unique_ptr<fs::path>, std::unique_ptr<fmt::ostream>, std::unique_ptr<std::ostream>, std::unique_ptr<FILE *>>::value;
 
     template <class T, FileTypes fileType>
-    inline constexpr bool enable_json_v = is_all_json_type(fileType) && is_any<T, nlohmann::json>::value &&
-                                          !is_any<T, std::string_view, std::string, char *>::value;
+    inline constexpr bool enable_json_v =
+        is_all_json_type(fileType) && is_any<T, nlohmann::json>::value && !is_any<T, std::string_view, std::string, char *>::value;
```
Comment (Myoldmopar) [inline src/EnergyPlus/FileSystem.hh]: Format only.

File: src/EnergyPlus/HeatBalanceKivaManager.cc [hunk 6]
```diff
@@ -839,7 +839,7 @@ bool KivaManager::setupKivaInstances(EnergyPlusData &state)
 
                     Real64 surfHeight = Surfaces(wl).get_average_height(state);
                     // round to avoid numerical precision differences
-                    surfHeight = std::round((surfHeight)*1000.0) / 1000.0;
+                    surfHeight = std::round((surfHeight) * 1000.0) / 1000.0;
```
Comment (Myoldmopar) [inline src/EnergyPlus/HeatBalanceKivaManager.cc]: Format only -- surprised that Clang Format ever allowed this before.

File: src/EnergyPlus/HeatPumpWaterToWaterCOOLING.hh [hunk 7]
```diff
@@ -128,11 +128,10 @@ namespace HeatPumpWaterToWaterCOOLING {
...
-              PowerLosses(0.0), LossFactor(0.0), HighPressCutoff(0.0), LowPressCutoff(0.0), IsOn(false),
-              MustRun(false), SourcePlantLoc{}, LoadPlantLoc{}, CondMassFlowIndex(0), Power(0.0), Energy(0.0), QLoad(0.0), QLoadEnergy(0.0),
-              QSource(0.0), QSourceEnergy(0.0), LoadSideWaterInletTemp(0.0), SourceSideWaterInletTemp(0.0), LoadSideWaterOutletTemp(0.0),
-              SourceSideWaterOutletTemp(0.0), Running(0), LoadSideWaterMassFlowRate(0.0), SourceSideWaterMassFlowRate(0.0), plantScanFlag(true),
-              beginEnvironFlag(true)
+              PowerLosses(0.0), LossFactor(0.0), HighPressCutoff(0.0), LowPressCutoff(0.0), IsOn(false), MustRun(false), SourcePlantLoc{},
+              LoadPlantLoc{}, CondMassFlowIndex(0), Power(0.0), Energy(0.0), QLoad(0.0), QLoadEnergy(0.0), QSource(0.0), QSourceEnergy(0.0),
+              LoadSideWaterInletTemp(0.0), SourceSideWaterInletTemp(0.0), LoadSideWaterOutletTemp(0.0), SourceSideWaterOutletTemp(0.0), Running(0),
+              LoadSideWaterMassFlowRate(0.0), SourceSideWaterMassFlowRate(0.0), plantScanFlag(true), beginEnvironFlag(true)
```
Comment (Myoldmopar) [inline src/EnergyPlus/HeatPumpWaterToWaterCOOLING.hh]: Format only, just changing the line cut off.

File: src/EnergyPlus/IOFiles.hh [hunk 8]
```diff
@@ -141,6 +141,9 @@ private:
     std::string_view spec_builder()
     {
         buffer.clear();
+        // This line seems to be the culprit with some array-bounds warnings
+        // during compilation on GCC 13.2.  I don't see how at the moment.
+        // I tried pragma-ing it away but it didn't help
```
Comment (Myoldmopar) [inline src/EnergyPlus/IOFiles.hh]: Some GCC warnings are emitted and mention this line.  Not really sure.  The other changes in this file are only format related.

File: src/EnergyPlus/PhaseChangeModeling/HysteresisModel.cc [hunk 9]
```diff
@@ -165,7 +165,7 @@ namespace HysteresisPhaseChange {
                     phaseChangeState = PhaseChangeStates::FREEZING;
                     this->enthNew =
                         this->getEnthalpy(updatedTempTDT, this->peakTempFreezing, this->deltaTempFreezingLow, this->deltaTempFreezingHigh);
-                } else if (this->enthNew < this->enthalpyF && this->enthNew > this->enthalpyM) {
+                } else if (this->enthNew > this->enthalpyM && this->enthNew < this->enthalpyF) {
```
Comment (Myoldmopar) [inline src/EnergyPlus/PhaseChangeModeling/HysteresisModel.cc:168]: Reordered some of the IF blocks so that Clang Format will stop trying to format them as template type argument specifiers.  Even Clang Format 17 is getting confused about it.

File: src/EnergyPlus/PluginManager.cc [hunk 10]
```diff
@@ -62,12 +62,10 @@ namespace EnergyPlus::PluginManagement {
 PluginTrendVariable::PluginTrendVariable(EnergyPlusData &state, std::string _name, int _numValues, int _indexOfPluginVariable)
     : name(std::move(_name)), numValues(_numValues), indexOfPluginVariable(_indexOfPluginVariable)
 {
-    // initialize the deque so it can be queried immediately, even with just zeroes
+    // initialize the deque, so it can be queried immediately, even with just zeroes
```
Comment (Myoldmopar) [inline src/EnergyPlus/PluginManager.cc:65]: Just collapsed the two loops into one.  Should be harmless

File: src/EnergyPlus/PluginManager.cc [hunk 11]
```diff
@@ -379,42 +377,27 @@ void PluginManager::setupOutputVariables([[maybe_unused]] EnergyPlusData &state)
...
-    // Now read all the actual plugins and interpret them
-    // IMPORTANT -- DO NOT CALL setup() UNTIL ALL INSTANCES ARE DONE
-    std::string const sPlugins = "PythonPlugin:Instance";
-    int pluginInstances = state.dataInputProcessing->inputProcessor->getNumObjectsFound(state, sPlugins);
-    if (pluginInstances == 0) {
-        return;
-    }
-
 #if LINK_WITH_PYTHON
+#if PY_MINOR_VERSION <= 10 // fix this
```
Comment (Myoldmopar) [inline src/EnergyPlus/PluginManager.cc:381]: Alright, so with newer Python we have to use the PyConfig form of interpreter initialization.  It's a great improvement, and I'm glad we're doing it, but I didn't want to break other versions.  I need to carefully go through and figure out which version of Python made the change, and see how to best massage this together.  For now, I am choosing Python 3.11 as the cut off to use the new version.  This **definitely** needs research/fix.

File: src/EnergyPlus/PluginManager.cc [hunk 12]
```diff
@@ -379,42 +377,27 @@ void PluginManager::setupOutputVariables([[maybe_unused]] EnergyPlusData &state)
...
-        Py_SetPath(ws.c_str());
-        Py_SetPythonHome(ws.c_str());
-    } else {
-        // TODO: Py_DecodeLocale shouldn't be called before Python is PreInitialized. Also, this should be replaced by PyConfig
-        wchar_t *a = Py_DecodeLocale(pathToPythonPackages.generic_string().c_str(), nullptr); // This allocates!
-        Py_SetPath(a);
-        Py_SetPythonHome(a);
-        PyMem_RawFree(a);
-    }
+#ifdef _WIN32
```
Comment (Myoldmopar) [inline src/EnergyPlus/PluginManager.cc:389]: So before this was:

```
if constexpr (std::is_same_v<typename fs::path::value_type, wchar_t>) {
```

Now it's

```
#ifdef _WIN32
```

The `fs::path::value_type` is a `wchar_t` on, and only on, Windows, so I think this is equivalent but much more readable.  I'm fine going back.  Also, maybe nowadays we can do better about handling them in a single function rather than having to be conditional, I'm not sure.  Questions for @jmarrec for sure.

File: src/EnergyPlus/PluginManager.cc [hunk 13]
```diff
@@ -428,6 +411,85 @@ PluginManager::PluginManager(EnergyPlusData &state) : eplusRunningViaPythonAPI(s
...
+        ShowFatalError(state, "Could not initialize program_name on PyConfig...weird");
+    }
+
+    status = PyConfig_Read(&config);
+    if (PyStatus_Exception(status)) {
+        ShowFatalError(state, "Could not read back the PyConfig...weird");
+    }
+
+#ifdef _WIN32
+    // TODO: Test this on Windows...
```
Comment (Myoldmopar) [inline src/EnergyPlus/PluginManager.cc:444]: Need to uncomment and test this on Windows with updated Python.

File: src/EnergyPlus/PluginManager.cc [hunk 14]
```diff
@@ -690,7 +752,7 @@ void PluginInstance::reportPythonError([[maybe_unused]] EnergyPlusData &state)
             return;
         }
 
-        unsigned long numVals = PyList_Size(pyth_val);
+        Py_ssize_t numVals = PyList_Size(pyth_val);
```
Comment (Myoldmopar) [inline src/EnergyPlus/PluginManager.cc:755]: There were several new warnings being emitted due to slight signed/unsigned mismatches and things like that.  I thought about ignoring, but it ended up only being a handful of changes and now everything is happy.

File: src/EnergyPlus/PluginManager.cc [hunk 15]
```diff
@@ -753,6 +815,7 @@ void PluginInstance::setup([[maybe_unused]] EnergyPlusData &state)
         EnergyPlus::ShowFatalError(state, "Python import error causes program termination");
     }
     PyObject *pModuleDict = PyModule_GetDict(this->pModule);
+    // printDict(pModuleDict);
```
Comment (Myoldmopar) [inline src/EnergyPlus/PluginManager.cc:818]: Remove this debug line.

File: third_party/ssc/shared/lib_battery_lifetime.h [hunk 16]
```diff
@@ -81,7 +81,7 @@ class lifetime_t {
 
     lifetime_t(const lifetime_t &rhs);
 
-    virtual lifetime_t &operator=(const lifetime_t &rhs);
+    lifetime_t &operator=(const lifetime_t &rhs);
```
Comment (Myoldmopar) [inline third_party/ssc/shared/lib_battery_lifetime.h:84]: I was getting a compiler warning about how the virtual operator was being hidden by derived classes.  By taking off the virtual, we should be getting the same behavior, but without the warning.  I could also mute the warning when we include this third party folder, I really don't care.  FYI @nmerket

File: third_party/ssc/shared/lib_battery_lifetime.h [hunk 17]
```diff
@@ -68,7 +68,7 @@ struct lifetime_state {
 
     lifetime_state(const std::shared_ptr<cycle_state>& cyc, const std::shared_ptr<calendar_state>& cal);
 
-    lifetime_state(const std::shared_ptr<lifetime_nmc_state>& nmc);
+    explicit lifetime_state(const std::shared_ptr<lifetime_nmc_state>& nmc);
```
Comment (Myoldmopar) [inline third_party/ssc/shared/lib_battery_lifetime.h:71]: Tiny warning about single argument constructors.  Would be happy to undo.

File: tst/EnergyPlus/unit/HVACVariableRefrigerantFlow.unit.cc [hunk 18]
```diff
@@ -2487,109 +2487,109 @@ TEST_F(EnergyPlusFixture, VRF_FluidTCtrl_VRFOU_Compressor)
     }
 
     // Run and Check: VRFOU_CompSpd
-    {//   Test the method VRFOU_CompSpd, which calculates the compressor speed at given
-     //   operational conditions to meet the evaporator or condenser capacity provided.
+    { //   Test the method VRFOU_CompSpd, which calculates the compressor speed at given
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/HVACVariableRefrigerantFlow.unit.cc]: Format changes only in this file.

**Review by Myoldmopar — COMMENTED** [review]
OK, walkthrough done, I already noticed a couple things I need to just take out from debugging steps.  I would love some thoughts on this from @jmarrec at least, and anyone else interested.  My goal is to avoid affecting anyone else's build at all, while also keeping the config/build/stuff as simple as possible.

File: cmake/CompilerFlags.cmake (hunk 2 above)
Comment (jmarrec) [reply cmake/CompilerFlags.cmake:114]: Define many/lots please? Obviously we're not fans of plain silencing a whole section of warnings, so just trying to gauge if feasible to pragma ignore some of them.

And yes, totally agreed that we should investigate if they are false positive.

(At minimum, I'd say this should be a conditional ignore)

File: src/EnergyPlus/PhaseChangeModeling/HysteresisModel.cc (hunk 9 above)
Comment (jmarrec) [reply src/EnergyPlus/PhaseChangeModeling/HysteresisModel.cc:168]: There should just be parenthesis around each condition, that'd improve readability too. And I think clang format will stop messing up.

```suggestion
                } else if ((this->enthNew < this->enthalpyF) && (this->enthNew > this->enthalpyM)) {
```

File: src/EnergyPlus/PluginManager.cc (hunk 11 above)
Comment (jmarrec) [reply src/EnergyPlus/PluginManager.cc:381]: have you tried compiling with python 3.8 (current) and the PyConfig stuff yet? Maybe it just works?

File: src/EnergyPlus/PluginManager.cc (hunk 12 above)
Comment (jmarrec) [reply src/EnergyPlus/PluginManager.cc:389]: So, in our general build cases (MSVC, macOS clang, Ubuntu gcc), the macro and the type_traits are indeed functionally equivalent. But I wasn't trying to just be "clever" and use type traits just because I could, I was actually trying to make it truly cross-platform.
What happens if someone tries to build with MinGW on windows for eg? I don't know. But it'll work with type_traits.
And the `if constexpr` does not make a runtime check, so no penalty.

TL;DR: I think this is a case where ignoring the warning is better.

File: src/EnergyPlus/PluginManager.cc [hunk 19]
```diff
@@ -428,6 +411,85 @@ PluginManager::PluginManager(EnergyPlusData &state) : eplusRunningViaPythonAPI(s
...
+    PyStatus status;
+
+    // first pre-config Python so that it can speak UTF-8
+    PyPreConfig preConfig;
+    PyPreConfig_InitPythonConfig(&preConfig);
+    preConfig.utf8_mode = 1;
+    status = Py_PreInitialize(&preConfig);
+    if (PyStatus_Exception(status)) {
+        ShowFatalError(state, "Could not pre-initialize Python to speak UTF-8...weird");
+    }
```
Comment (jmarrec) [inline src/EnergyPlus/PluginManager.cc:427]: Nice. Perhaps we should improve the error message here, maybe it'll save us some trouble if someone somehow runs into that error and we can't reproduce...

I think this formatter should do. I'll test locally.

```c+++
template <> struct fmt::formatter<PyStatus>
{
    // parse is inherited from formatter<string_view>.
    constexpr auto parse(format_parse_context &ctx) -> format_parse_context::iterator
    {
        return ctx.begin();
    }

    auto format(const PyStatus &status, format_context &ctx) const -> format_context::iterator
    {
        if (!PyStatus_Exception(status)) {
            return ctx.out();
        }
        if (PyStatus_IsExit(status)) {
            return fmt::format_to(ctx.out(), "Exited with code {}", status.exitcode);
        }
        if (PyStatus_IsError(status)) {
            auto it = ctx.out();
            it = fmt::format_to(it, "Fatal Python error: ");
            if (status.func) {
                it = fmt::format_to(it, "{}: ", status.func);
            }
            it = fmt::format_to(it, "{}", status.err_msg);
            return it;
        }
        return ctx.out();
    }
};
```

Usage: `ShowFatalError(state, fmt::format("Could not pre-initialize Python to speak UTF-8... {}", status));`

File: src/EnergyPlus/PluginManager.cc [hunk 20]
```diff
@@ -1180,16 +1241,17 @@ void PluginManager::addGlobalVariable([[maybe_unused]] EnergyPlusData &state, [[
 int PluginManager::getGlobalVariableHandle(EnergyPlusData &state, const std::string &name, bool const suppress_warning)
 { // note zero is a valid handle
     std::string const varNameUC = EnergyPlus::Util::makeUPPER(name);
-    auto const it = std::find(state.dataPluginManager->globalVariableNames.begin(), state.dataPluginManager->globalVariableNames.end(), varNameUC);
-    if (it != state.dataPluginManager->globalVariableNames.end()) {
-        return std::distance(state.dataPluginManager->globalVariableNames.begin(), it);
+    auto const &gVarNames = state.dataPluginManager->globalVariableNames;
+    auto const it = std::find(gVarNames.begin(), gVarNames.end(), varNameUC);
+    if (it != gVarNames.end()) {
+        return (int)std::distance(gVarNames.begin(), it);
```
Comment (jmarrec) [inline src/EnergyPlus/PluginManager.cc:1247]: C'MON. C style cast?

File: src/EnergyPlus/PluginManager.cc [hunk 21]
```diff
@@ -428,6 +411,85 @@ PluginManager::PluginManager(EnergyPlusData &state) : eplusRunningViaPythonAPI(s
...
+    // we'll need the program directory for a few things so get it once here at the top and sanitize it
+    fs::path programDir;
+    if (state.dataGlobal->installRootOverride) {
+        programDir = state.dataStrGlobals->exeDirectoryPath;
+    } else {
+        programDir = FileSystem::getParentDirectoryPath(FileSystem::getAbsolutePath(FileSystem::getProgramPath()));
+    }
+    fs::path const pathToPythonPackages = programDir / "python_standard_lib";
+
+#ifdef LINK_WITH_PYTHON
```
Comment (jmarrec) [inline src/EnergyPlus/PluginManager.cc:491]: Wrong position, and wrong macro (shoudl be if, not ifdef)

File: src/EnergyPlus/PluginManager.cc (hunk 11 above)
Comment (jmarrec) [reply src/EnergyPlus/PluginManager.cc:381]: I'm actually a strong proponent of making PyConfig changes into its own PR so we can just test wiht python 3.8. I'm doing that locally as we speak.

**Review by jmarrec — CHANGES_REQUESTED** [review]

**Comment by jmarrec** [comment]
@Myoldmopar I extracted the PluginManager changes into its own PR at https://github.com/NREL/EnergyPlus/pull/10342, enabling them for all python versions (including 3.8, our currently supported Python... PyConfig was added in 3.8). It builds and tests perfect on my mac.

**Comment by Myoldmopar** [comment]
So @jmarrec took the work here and split it into a couple PRs, and enhanced them each there.  This can be closed, thanks @jmarrec
